File: NRDB/NROutlineLabel.m
//
//  NROutlineLabel.m
//  NRDB
//
//  Created by Gereon Steffens on 24.05.14.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "NROutlineLabel.h"

@implementation NROutlineLabel

- (id) initWithFrame:(CGRect)frame
{
    if ((self = [super initWithFrame:frame]))
    {
        [self awakeFromNib];
    }
    return self;
}

- (void) awakeFromNib
{
    self.outlineWidth = 2;
    self.outlineColor = [UIColor whiteColor];
}

- (void)drawTextInRect:(CGRect)rect {
    
    CGSize shadowOffset = self.shadowOffset;
    UIColor *textColor = self.textColor;
    
    CGContextRef c = UIGraphicsGetCurrentContext();
    CGContextSetLineWidth(c, self.outlineWidth);
    CGContextSetLineJoin(c, kCGLineJoinRound);
    
    CGContextSetTextDrawingMode(c, kCGTextStroke);
    self.textColor = self.outlineColor;
    [super drawTextInRect:rect];
    
    CGContextSetTextDrawingMode(c, kCGTextFill);
    self.textColor = textColor;
    self.shadowOffset = CGSizeMake(0, 0);
    [super drawTextInRect:rect];
    
    self.shadowOffset = shadowOffset;
}

@end

File: NetDeck/NROutlineLabel.swift
//
//  NROutlineLabel.swift
//  NetDeck
//
//  Created by Gereon Steffens on 24.04.16.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

import UIKit

class NROutlineLabel: UILabel {
    fileprivate let outlineWidth: CGFloat = 2.0
    fileprivate let outlineColor = UIColor.white
    
    override func drawText(in rect: CGRect) {
        let textColor = self.textColor
        let shadowOffset = self.shadowOffset
        
        let ctx = UIGraphicsGetCurrentContext()
        ctx?.setLineWidth(self.outlineWidth)
        ctx?.setLineJoin(.round)
        ctx?.setTextDrawingMode(.stroke)
        self.textColor = self.outlineColor
        super.drawText(in: rect)
        
        ctx?.setTextDrawingMode(.fill)
        self.textColor = textColor
        self.shadowOffset = CGSize.zero
        super.drawText(in: rect)
        
        self.shadowOffset = shadowOffset
    }
}
